controllers/nodemaintenance_controller.go

```diff
@@ -109,112 +113,123 @@ func (r *NodeMaintenanceReconciler) Reconcile(ctx context.Context, req ctrl.Requ
+	err = r.initMaintenanceStatus(nm)
 	if err != nil {
 		r.logger.Error(err, "Failed to update NodeMaintenance with \"Running\" status")
-		return r.onReconcileError(instance, err)
+		return r.onReconcileError(nm, err)
 	}
 
-	nodeName := instance.Spec.NodeName
+	nodeName := nm.Spec.NodeName
 
-	r.logger.Info("Applying maintenance mode", "node", nodeName, "reason", instance.Spec.Reason)
+	r.logger.Info("Applying maintenance mode", "node", nodeName, "reason", nm.Spec.Reason)
 	node, err := r.fetchNode(nodeName)
 	if err != nil {
-		return r.onReconcileError(instance, err)
+		return r.onReconcileError(nm, err)
 	}
 
-	r.setOwnerRefToNode(instance, node)
+	r.setOwnerRefToNode(nm, node)
 
 	updateOwnedLeaseFailed, err := r.obtainLease(node)
 	if err != nil && updateOwnedLeaseFailed {
-		instance.Status.ErrorOnLeaseCount += 1
-		if instance.Status.ErrorOnLeaseCount > maxAllowedErrorToUpdateOwnedLease {
+		nm.Status.ErrorOnLeaseCount += 1
+		if nm.Status.ErrorOnLeaseCount > maxAllowedErrorToUpdateOwnedLease {
 			r.logger.Info("can't extend owned lease. uncordon for now")
 
 			// Uncordon the node
 			err = r.stopNodeMaintenanceImp(ctx, node)
 			if err != nil {
-				return r.onReconcileError(instance, fmt.Errorf("Failed to uncordon upon failure to obtain owned lease : %v ", err))
+				return r.onReconcileError(nm, fmt.Errorf("failed to uncordon upon failure to obtain owned lease : %v ", err))
 			}
-			instance.Status.Phase = nodemaintenancev1beta1.MaintenanceFailed
+			// maintenance has failed - node was uncordon and under maintenance mode
+			utils.WarningEvent(r.Recorder, nm, utils.EventReasonFailedMaintenance, utils.EventMessageFailedMaintenance)
+			nm.Status.Phase = nodemaintenancev1beta1.MaintenanceFailed
 		}
-		return r.onReconcileError(instance, fmt.Errorf("failed to extend lease owned by us : %v errorOnLeaseCount %d", err, instance.Status.ErrorOnLeaseCount))
+		return r.onReconcileError(nm, fmt.Errorf("failed to extend lease owned by us : %v errorOnLeaseCount %d", err, nm.Status.ErrorOnLeaseCount))
 	}
 	if err != nil {
-		instance.Status.ErrorOnLeaseCount = 0
-		return r.onReconcileError(instance, err)
+		nm.Status.ErrorOnLeaseCount = 0
+		return r.onReconcileError(nm, err)
 	} else {
-		if instance.Status.Phase != nodemaintenancev1beta1.MaintenanceRunning || instance.Status.ErrorOnLeaseCount != 0 {
-			instance.Status.Phase = nodemaintenancev1beta1.MaintenanceRunning
-			instance.Status.ErrorOnLeaseCount = 0
+		if nm.Status.Phase != nodemaintenancev1beta1.MaintenanceRunning || nm.Status.ErrorOnLeaseCount != 0 {
+			// Another chance to evict pods - clear ErrorOnLeaseCount and try again to put the node under maintenance
+			utils.NormalEvent(r.Recorder, nm, utils.EventReasonEvictingPods, utils.EventMessageEvictingPods)
+			nm.Status.Phase = nodemaintenancev1beta1.MaintenanceRunning
+			nm.Status.ErrorOnLeaseCount = 0
 
 		}
 	}
-	//Add exclude from remediation label
+	// Add exclude from remediation label
```

Comment:
nit: considering the name of the function is quite clear, I'd consider removing the comment altogether, but up to you